Sources/_SwiftSyntaxTestSupport/IncrementalParseTestUtils.swift

```diff
@@ -0,0 +1,111 @@
+public func assertIncrementalParse(
+  _ source: String,
+  reusedNodes expectedReusedNodes: [ReusedNodeSpec] = [],
+  file: StaticString = #file,
+  line: UInt = #line
+) {
+  let (concurrentEdits, originalSource, editedSource) = getEditsAndSources(source)
+  
+  let originalString = String(originalSource)
+  let editedString = String(editedSource)
+  
+  let originalTree = Parser.parse(source: originalString)
+
+  let reusedNodesCollector = IncrementalParseReusedNodeCollector()
+  let transition = IncrementalParseTransition(previousTree: originalTree, edits: concurrentEdits, reusedNodeDelegate: reusedNodesCollector)
+
+  let newTree = Parser.parse(source: String(editedSource))
+  let incrementallyParsedNewTree = Parser.parse(source: editedString, parseTransition: transition)
+
+  // Round-trip
+  assertStringsEqualWithDiff(
+    editedString,
+    "\(incrementallyParsedNewTree)",
+    additionalInfo: """
+      Source failed to round-trip when parsing incrementally
+      
+      Actual syntax tree:
+      \(incrementallyParsedNewTree.debugDescription)
+      """,
+    file: file,
+    line: line
+  )
+
+  // Substructure
+  let subtreeMatcher = SubtreeMatcher(Syntax(incrementallyParsedNewTree), markers: [:])
+  do {
+    try subtreeMatcher.assertSameStructure(Syntax(newTree), includeTrivia: true, file: file, line: line)
+  } catch {
+    XCTFail("Matching for a subtree failed with error: \(error)", file: file, line: line)
+  }
+
+  // Re-used nodes
+  // The range of node should ignore Trivia
+  XCTAssertEqual(reusedNodesCollector.rangeAndNodes.count, expectedReusedNodes.count, file: file, line: line)
+  
+  for targetNode in expectedReusedNodes {
+    guard let range = getByteSourceRange(for: targetNode.sourceString, in: originalString) else {
+      XCTFail("Fail to find \(targetNode.sourceString) in \(originalString), please check test setting.")
+      return
+    }
+    
+    guard let reusedNode = reusedNodesCollector.rangeAndNodes.first(where: { $0.0 == range })?.1 else {
+      XCTFail(
+        """
+        Fail to match the range of \(targetNode.sourceString) in:
+        \(reusedNodesCollector.rangeAndNodes.map({"\($0.0): \($0.1.description)"}).joined(separator: "\n"))
+        """,
+        file: file, line: line
+      )
+      return
+    }
+    
+    XCTAssertEqual(targetNode.kind, reusedNode.kind, file: file, line: line)
+  }
+}
+
+private func getByteSourceRange(for subString: String, in sourceString: String) -> ByteSourceRange? {
+  
+  if let range = sourceString.range(of: subString) {
+    
+    return ByteSourceRange(
+      offset: sourceString.utf8.distance(from: sourceString.startIndex, to: range.lowerBound),
+      length: sourceString.utf8.distance(from: range.lowerBound, to: range.upperBound))
+  }
+  return nil
+}
+
+public struct ReusedNodeSpec {
+  public let sourceString: String
+  
+  public let kind: SyntaxKind
+  
+  public init(
+    sourceString: String,
+    kind: SyntaxKind
+  ) {
```

Comment:
Just an idea to make the tests a little bit more readable: I’d make the first parameter unlabeled. But that’s very much personal preference.